Add compression for uploaded documentation

jyn514 opened the pull request:
This needs some tests, but otherwise I think it should be fine as is, it was super simple after #643 :)

This uses `zstd 5`, but it's very easy to change the compression now and reasonably easy to change it after it's merged (we just have to change `compression` from a boolean into an enum and continue supporting `zstd`).

This compresses transparently, so that calling `backend.get()` automatically decompresses files as needed. If the files were not compressed before uploading, they are not decompressed when downloading.

Closes https://github.com/rust-lang/docs.rs/issues/379

cc @namibj
r? @pietroalbini

jyn514 commented:
Oh, I just realized this has the wrong content-types on s3: It uses the original content types instead of `zstd` or something like that. I'll add the original content type as a field in the metadata.

jyn514 commented:
Actually that seems like a lot of work for no benefit, I don't know that we'll ever actually use the `zstd` encoding for anything.

jyn514 commented:
Some benchmarks:
```
Benchmarking compress regex html: Collecting 100 samples in estimated 7.5520 s (10k i                                                                                     compress regex html     time:   [676.22 us 725.12 us 779.81 us]
Found 16 outliers among 100 measurements (16.00%)
  16 (16.00%) high severe

Benchmarking decompress regex html: Collecting 100 samples in estimated 5.1476 s (91k                                                                                     decompress regex html   time:   [55.181 us 57.623 us 60.467 us]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

Kixiron commented on src/db/migrate.rs:
```diff
@@ -325,6 +325,17 @@ pub fn migrate(version: Option<Version>, conn: &Connection) -> CratesfyiResult<(
                 ALTER TABLE releases ALTER COLUMN doc_targets DROP NOT NULL;
             "
         ),
+        migration!(
+            context,
+            // version
+            13,
+            // description
+            "Add a field for compression",
+            // upgrade query
+            "ALTER TABLE files ADD COLUMN compressed BOOLEAN NOT NULL DEFAULT false;",
+            // downgrade query
+            "ALTER TABLE files DROP COLUMN compressed;",
+        ),
```
I still think that storing this in the s3 header is the way to go

pietroalbini replied:
For S3 yes, this is just for the database storage backend.

Nemo157 commented on src/storage/s3.rs:
```diff
@@ -51,19 +51,25 @@ impl<'a> S3Backend<'a> {
+
         let mut attempts = 0;
+        let mut compressed = HashMap::with_capacity(1);
+        compressed.insert("compressed".into(), "true".into());
```
Rather than just compressed or not, this should store the encoding used, so that it’s possible to move to different encodings in the future.

Nemo157 replied:
In fact, it might make a lot of sense to use the `Content-Encoding` header instead if that’s something that S3 retains.

jyn514 replied:
`Content-Encoding` is a great idea, I didn't know it existed!

pietroalbini replied:
The `rusoto_s3` field is this: https://docs.rs/rusoto_s3/0.43.0/rusoto_s3/struct.PutObjectRequest.html#structfield.content_encoding

jyn514 replied:
Added in https://github.com/rust-lang/docs.rs/pull/780/commits/f13a08783b97d6af93e6603e1a8298d1168af09f

pietroalbini commented:
The implementation looks good to me!

I agree with @Nemo157 that we should store `{algorithm}` instead of just `compressed`, and that we should use S3's native `Content-Encoding` to store it instead of a custom metadata.

Also, based on https://github.com/rust-lang/docs.rs/issues/379#issuecomment-512834913 the two options I was considering were `zstd 9` or `brotli 5` leaning torwards brotli as it uses 10% less disk space in that benchmark. You're using `zstd 5` in the PR.

Nemo157 commented:
The other option that @namibj was investigating was zstd with a custom dictionary optimized for rustdoc output, I can probably adapt the linked benchmarking script to test that as well with a dictionary created from the docs I have.

jyn514 commented:
Oops, the benchmarks before got cut off. Here are some taken with a desktop instead of a laptop:
```
compress regex html     time:   [715.28 us 715.74 us 716.27 us]                                
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

decompress regex html   time:   [37.410 us 37.438 us 37.470 us]                                   
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
```

Nemo157 commented:
Another idea might be to store a set of all the encodings used for a particular release into the `release` table, so we can easily lookup basic stats on compression usage without needing to go to S3. (A set to allow for different dictionaries based on filetype to be recorded).

jyn514 commented:
> store a set of all the encodings used for a particular release into the release table

This sounds like we'd only use it for metrics? I'm not opposed in principle but storing the same data in two places risks it getting out of sync.

Kixiron commented:
Metrics don’t need to be a source of complexity, we can just as easily have an `IntCounter` that we increment with the total bytes saved by each build (plus additional if we’d like)

Nemo157 commented:
It's not really for metrics like what goes in grafana, more for future visibility into usage distribution. Especially with if we migrate dictionaries in the future knowing how many releases use which dictionaries could be useful to determine whether it seems feasible to do something like migrate all instances of a certain dictionary to a new one to remove it from the application.

Given that releases are basically read-only currently I don't think there's much chance of this data getting out of sync.

Kixiron commented:
How would that interact with us deleting docs in the future (if we ever decided to do that)?

jyn514 commented:
> How would that interact with us deleting docs in the future (if we ever decided to do that)?

No effect, we'd just delete the rows out of the database as well.

jyn514 commented:
> Another idea might be to store a set of all the encodings used for a particular release into the release table, so we can easily lookup basic stats on compression usage without needing to go to S3. (A set to allow for different dictionaries based on filetype to be recorded).

@Nemo157 a possible issue with this is it prevents using different compression algorithms for different files within a release. Does that sound worth it? Otherwise we'd have to store metadata for each file individually in the database which sort of the defeats the point of an S3 bucket.

Nemo157 commented:
That's why I mention a set of encodings, we don't need to know exactly which files use which encoding, just that having this set of encodings is enough to access all files in the release (and if we want to know in detail then we would have to query the metadata on all the files).

jyn514 commented:
I added the set of compression algorithms used as a many-to-many-table in the database, with a unique constraint so we don't end up with duplicates by accident.

I'm open to suggestions for more tests but otherwise I think I addressed all the comments.

jyn514 commented:
As to using a custom zstd dictionary: it sounds like the rust `zstd` crate needs some fairly major changes to make that performant, I don't want to delay having any compression at all on that. If we want to use brotli instead for the storage that sounds reasonable but otherwise I think `zstd 9` is fine.

Nemo157 commented on src/db/migrate.rs:
```diff
@@ -340,6 +340,33 @@ pub fn migrate(version: Option<Version>, conn: &Connection) -> CratesfyiResult<(
+            );",
+            // downgrade query
+            "DROP TABLE compression_rels;
+             DROP TABLE compression;",
```
This seems like it would fail given `files.compression` still exists.

jyn514 replied:
Good catch, fixed.

Nemo157 commented:
I think it definitely makes sense to start with `zstd` now and add custom dictionary support later. Going with brotli now and moving to zstd custom dictionary later would require keeping both libraries around indefinitely.

Doing it in two steps also gives us the chance to see any performance impacts of each step.

One last thought I had was whether there should be a config var to disable compression, so if there are any issues in production that can just be toggled rather than having to do a rollback.

jyn514 commented:
> One last thought I had was whether there should be a config var to disable compression, so if there are any issues in production that can just be toggled rather than having to do a rollback.

It doesn't take very long to revert, about 5 minutes. In general I like the idea of having feature gates instead of "it's compiled in or it's not" but I'm a little wary of toggling features with environment variables, I'd rather come up with something more principled.

pietroalbini commented on src/storage/mod.rs:
```diff
@@ -6,20 +6,47 @@ pub(crate) use self::s3::S3Backend;
 
+pub type CompressionAlgorithms = HashSet<CompressionAlgorithm>;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CompressionAlgorithm {
+    Zstd,
+}
+
+impl std::str::FromStr for CompressionAlgorithm {
+    type Err = ();
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "zstd" => Ok(CompressionAlgorithm::Zstd),
+            _ => Err(()),
+        }
+    }
+}
+
+impl fmt::Display for CompressionAlgorithm {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CompressionAlgorithm::Zstd => write!(f, "zstd"),
+        }
+    }
+}
```
I'm not convinced this is the best way to represent the compression algorithm.

A thing I usually strive for when creating this kind of code is making sure there is only a single place where a new variant needs to be added: this way a new contributor can just add the algorithm to the enum and follow the compiler errors to implement full support for it. This advantage is not present here, as the contributor also needs to add the algorithm name in the database through a migration, which is not intuitive nor documented anywhere.

We *could* implement some code that synchronizes the `compression` table when the application starts, but I think we can implement this more cleanly. If we remove the string representation and manually set each ID in the discriminant ([with the help of a macro](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b4f6bc42c2c6caff6377b5ec9f282318)) we'd remove a lot of complexity and make it easier for new contributors to add new variants.

jyn514 replied:
The macro is not necessary, you can use `algorithm as i32` instead. Other than that I agree, storing the id would be better than storing a string.

pietroalbini replied:
While that works, there is no builtin conversion from `i32` to `CompressionAlgorithm`.

jyn514 replied:
For that, is the duplication so bad? I think `impl TryFrom<i32> for CompressionAlgorithm` would work without obscuring what's happening: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=acde692246a13cddc897fa0e1927d004

jyn514 replied:
If you're worried that doesn't give a compile error if you forget to update the `match`, we could use https://crates.io/crates/num_enum.

pietroalbini replied:
Yeah my worry is it doesn't give a compilation error.

jyn514 replied:
I added a test which will fail to compile if you don't add an exhaustive match for the new variant, hopefully that's close enough. I'm not sure what this should look like for S3, if we just have a number for the content-encoding it's not super descriptive, but if it's a string we're back to the same issue of needing to update the code in many places.

Nemo157 replied:
You could use `strum` to generate the `Display` and `FromStr` impls (though that doesn’t help with syncing with the database).

jyn514 replied:
As another alternative I could go with 'update code in many places' and add a test that `FromStr` is exhaustive.

pietroalbini replied:
I think it's fine for S3 to use the string, but I think going with a custom macro is still the best option, especially considered how many (duplicated) impls we have.

jyn514 replied:
Ok, I went with the macro-based approach for extensibility. We can always change it later if need be.

jyn514 commented:
Let me know if there are any more changes that need to be made, I think I addressed all the review comments.

jyn514 commented:
I just found that this will prevent us from updating builds if we rebuild a crate:

```
thread 'main' panicked at 'Building documentation failed: Error(Db(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("23505"), message: "duplicate key value violates unique constraint \"compression_rels_release_algorithm_key\"", detail: Some("Key (release, algorithm)=(1, 0) already exists."), hint: None, position: None, where_: None, schema: Some("public"), table: Some("compression_rels"), column: None, datatype: None, constraint: Some("compression_rels_release_algorithm_key"), file: Some("nbtinsert.c"), line: Some(434), routine: Some("_bt_check_unique") }))', src/bin/cratesfyi.rs:321:21
```

I think I need to add an `ON CONFLICT DO UPDATE` clause.
